data/org.freedesktop.portal.Notification.xml

```diff
@@ -116,8 +166,211 @@
           * ``target`` (``v``)
 
             Target parameter to send along when activating the action.
+
+          * ``purpose`` (``s``)
+
+            The ``purpose`` of the button. This information may be used by the
+            notification server to style the button specially.
+
+            The standardized hints are defined as part of a ``content-type``.
+            Depending on the ``purpose`` other fields of the button may be ignored.
+            If the server doesn't understand the ``purpose` it will be
+            ignored and the button will be shown as a normal button.
+
+            Additional purposes may be defined by notification servers using
+            ``x-vendor`` prefix e.g. ``x-gnome.class.specific``
+
+        * ``actions`` (``aa{sv}``)
+
+          Array of actions that the notification server will invoke
+          automatically in specific cases. These actions may be called never or even multiple times.
+
+          Invoked actions will contain as parameter a vardict ``a{sv}`` containing the
+          ``target` and the response of the action.
+
+          The format for a serialized action is a vardict with the following supported keys:
+
+          * ``action`` (``s``)
+
+            Name of the action. Mandatory.
+
+            This action needs to be exported by the application.
+
+          * ``purpose`` (``s``)
+            The purpose of the action. Mandatory.
+
+            * ``system.custom-alert``:
+
+              This action will be called by the notification server
+              whenever the notification is shown.
+
+              The purpose of this action is to allow apps to use custom methods of
+              notifying the user, for example, to play audio from a special
+              source like a streaming service or a radio station.
+
+            * ``user.reply-with-text``:
+
+              Inline replies for instant messaging apps.
+              The user-provided text will be added to the response vardict as ``reply-text`` (``s``);
+
+            Additional purposes may be defined by notification servers using
+            ``x-vendor`` prefix e.g. ``x-gnome.class.specific``
+
+          * ``target`` (``v``)
+
+            Target parameter to send along when activating the action.
+
+            This target will be added to the response vardict as ``target`` (``v``);
+
+
+        * ``desktop-file-id`` (``s``)
+
+          An optional ``desktop-file-id``.
+
+          If set, this must be a valid
+          `Desktop File ID <https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html#desktop-file-id>`_
+          with ``.desktop`` as a suffix.
+
+          For sandboxed apps, ``desktop-file-id``
+          must have the app ID followed by a period as a prefix.
```

Comment:
This sounds like sandboxed apps aren't allowed to specify their app ID as entire desktop file ID rather than a prefix but I hope that's not what is really meant?

Reply:
sandbox apps don't need to specify a desktop-file-id if it's the same as the app-id, since it will be picked up by the portal automatically.

Reply:
Well, they may not be aware whether they're in sandbox or not and just always use the same code.

Reply:
Right, but if the app sets the app id correctly it works also outside of flatpak. (Iirc on with systemd). Bit this is a valid point, i my need to dive deeper. I guess we can allow to set exactly the app id.

Reply:
> Right, but if the app sets the app id correctly it works also outside of flatpak. (Iirc on with systemd). 

Not sure what you mean? Currently portal gets app ID from the cgroups and the cgroup gets inherited when application is started by some other application as a URL handler (xdg-open), via terminal emulator or by clicking on the binary in file manager. I.e. it's a matter of how application is started rather than how it sets the app ID.